File: tools/benchmark.py
#!/usr/bin/env python3
"""
Benchmark detection models on the Raspberry Pi.

Measures pure inference FPS on a synthetic frame, which
makes model-to-model comparison fair and camera-independent.

Usage (on the Pi):
    python tools/benchmark.py --model models/yolov8n_int8_320.tflite
    python tools/benchmark.py --model models/yolov8n_int8_320.tflite models/ssd_mobilenet_v2_coco_int8_320.tflite --frames 100
"""

import argparse
import sys
import time
from pathlib import Path

import cv2
import numpy as np

PROJECT_DIRECTORY = Path(__file__).resolve().parent.parent

sys.path.insert(
    0,
    str(PROJECT_DIRECTORY / "src")
)


def build_detector(model_path, threads):
    """Return the right detector for the given model file."""

    from tflite_detector import TFLiteDetector

    return TFLiteDetector(
        model_path,
        threads=threads
    )


def benchmark(model_path, threads, frames):
    """Time inference on a batch of frames."""

    try:

        detector = build_detector(
            model_path,
            threads
        )

    except ImportError as error:

        print(
            f"{model_path.name:45s} SKIP "
            f"(install tflite-runtime on the Pi: {error})"
        )

        return 0

    # Synthetic 480p frame (typical camera resolution).

    frame = np.zeros(
        (480, 640, 3),
        dtype=np.uint8
    )

    # Warm up.

    detector.detect_person(
        frame,
        confidence_threshold=0.5
    )

    start = time.perf_counter()

    for _ in range(frames):

        detector.detect_person(
            frame,
            confidence_threshold=0.5
        )

    elapsed = time.perf_counter() - start

    fps = frames / elapsed

    print(
        f"{model_path.name:45s} "
        f"{1000 * elapsed / frames:8.1f} ms "
        f"{fps:6.2f} FPS"
    )

    return fps


def main():

    parser = argparse.ArgumentParser(
        description="Benchmark int8 TFLite models on the Pi."
    )

    parser.add_argument(
        "--model",
        nargs="+",
        default=[
            "models/ssd_mobilenet_v2_coco_int8_300.tflite",
            "models/efficientdet_lite0_coco_int8_320.tflite",
        ],
        help="Model file(s) to benchmark."
    )

    parser.add_argument(
        "--frames",
        type=int,
        default=100,
        help="Frames to time per model (default: 100)."
    )

    parser.add_argument(
        "--threads",
        type=int,
        default=4,
        help="Inference threads (default: 4)."
    )

    args = parser.parse_args()

    print(
        f"Benchmarking with {args.threads} thread(s), "
        f"{args.frames} frames each.\n"
    )

    best_model = None

    best_fps = 0

    for model in args.model:

        path = Path(model)

        if not path.exists():
            print(f"Skip (not found): {model}")
            continue

        fps = benchmark(
            path,
            threads=args.threads,
            frames=args.frames
        )

        if fps > best_fps:

            best_fps = fps

            best_model = model

    print("\nFastest model:")
    print(f"  {best_model} @ {best_fps:.1f} FPS")


if __name__ == "__main__":
    main()

File: cpp/tools/benchmark.cpp
// Benchmark detection models on the Raspberry Pi.
//
// Mirrors tools/benchmark.py: measures pure inference FPS on a
// synthetic frame, making model-to-model comparison fair and
// camera-independent.
//
// Usage (on the Pi):
//     ./benchmark --model models/ssd_mobilenet_v2_coco_int8_300.tflite
//     ./benchmark --model m1.tflite m2.tflite --frames 100

#include <chrono>
#include <filesystem>
#include <iomanip>
#include <iostream>
#include <memory>
#include <string>
#include <vector>

#include <opencv2/core.hpp>

#include "tflite_detector.hpp"

namespace {

namespace fs = std::filesystem;

struct Args {
    std::vector<std::string> models;
    int frames = 100;
    int threads = 4;
};

Args parseArgs(int argc, char** argv) {
    Args args;
    for (int i = 1; i < argc; ++i) {
        const std::string flag = argv[i];
        auto next = [&]() -> std::string {
            if (i + 1 < argc) {
                return argv[++i];
            }
            return std::string();
        };

        if (flag == "--frames") {
            args.frames = std::stoi(next());
        } else if (flag == "--threads") {
            args.threads = std::stoi(next());
        } else if (flag == "--model") {
            args.models.push_back(next());
        } else {
            // Bare positional arguments are treated as model paths.
            args.models.push_back(flag);
        }
    }
    return args;
}

/** Time inference on a batch of frames. Returns FPS (0 when skipped). */
double benchmark(const fs::path& model_path, int threads, int frames) {
    std::unique_ptr<tp::TFLiteDetector> detector;
    try {
        detector = std::make_unique<tp::TFLiteDetector>(model_path.string(), threads);
    } catch (const std::exception& error) {
        std::cout << std::left << std::setw(45)
                  << model_path.filename().string()
                  << " SKIP (TFLite unavailable: " << error.what() << ")\n";
        return 0.0;
    }

    // Synthetic 480p frame (typical camera resolution).
    cv::Mat frame = cv::Mat::zeros(480, 640, CV_8UC3);

    // Warm up.
    detector->detectPerson(frame, 0.5f);

    const auto start = std::chrono::steady_clock::now();
    for (int i = 0; i < frames; ++i) {
        detector->detectPerson(frame, 0.5f);
    }
    const auto end = std::chrono::steady_clock::now();

    const double elapsed =
        std::chrono::duration<double>(end - start).count();
    const double ms = 1000.0 * elapsed / frames;
    const double fps = frames / elapsed;

    std::cout << std::left << std::setw(45)
              << model_path.filename().string() << " "
              << std::fixed << std::setprecision(1) << std::setw(8) << ms
              << " ms " << std::setprecision(2) << std::setw(6) << fps
              << " FPS\n";
    return fps;
}

}  // namespace

int main(int argc, char** argv) {
    const Args args = parseArgs(argc, argv);

    std::vector<std::string> models = args.models;
    if (models.empty()) {
        models = {
            "models/ssd_mobilenet_v2_coco_int8_300.tflite",
            "models/efficientdet_lite0_coco_int8_320.tflite",
        };
    }

    std::cout << "Benchmarking with " << args.threads
              << " thread(s), " << args.frames << " frames each.\n\n";

    std::string best_model;
    double best_fps = 0.0;

    for (const auto& model : models) {
        const fs::path path = model;
        if (!fs::exists(path)) {
            std::cout << "Skip (not found): " << model << "\n";
            continue;
        }

        const double fps = benchmark(path, args.threads, args.frames);
        if (fps > best_fps) {
            best_fps = fps;
            best_model = model;
        }
    }

    std::cout << "\nFastest model:\n";
    std::cout << "  " << best_model << " @ " << best_fps << " FPS\n";

    return 0;
}